lightning-background-processor/src/lib.rs

```diff
@@ -263,30 +291,55 @@ mod tests {
+	macro_rules! end_open_channel {
+		($node_a: expr, $node_b: expr, $temporary_channel_id: expr, $tx: expr) => {{
+			$node_a.node.funding_transaction_generated(&$temporary_channel_id, $tx.clone()).unwrap();
 			$node_b.node.handle_funding_created(&$node_a.node.get_our_node_id(), &get_event_msg!($node_a, MessageSendEvent::SendFundingCreated, $node_b.node.get_our_node_id()));
 			$node_a.node.handle_funding_signed(&$node_b.node.get_our_node_id(), &get_event_msg!($node_b, MessageSendEvent::SendFundingSigned, $node_a.node.get_our_node_id()));
-			tx
 		}}
 	}
 
+	fn confirm_transaction(node: &mut Node, tx: &Transaction) {
+		for i in 1..=ANTI_REORG_DELAY {
+			let prev_blockhash = node.best_block.block_hash();
+			let height = node.best_block.height() + 1;
+			let header = BlockHeader { version: 0x20000000, prev_blockhash, merkle_root: Default::default(), time: height, bits: 42, nonce: 42 };
+			let txdata = vec![(0, tx)];
+			node.best_block = BestBlock::new(header.block_hash(), height);
+			match i {
+				1 => {
+					node.node.transactions_confirmed(&header, &txdata, height);
+					node.chain_monitor.transactions_confirmed(&header, &txdata, height);
+				},
+				ANTI_REORG_DELAY => {
+					node.node.best_block_updated(&header, height);
+					node.chain_monitor.best_block_updated(&header, height);
+				},
+				_ => {},
+			}
+		}
+	}
+
 	#[test]
 	fn test_background_processor() {
 		// Test that when a new channel is created, the ChannelManager needs to be re-persisted with
 		// updates. Also test that when new updates are available, the manager signals that it needs
 		// re-persistence and is successfully re-persisted.
 		let nodes = create_nodes(2, "test_background_processor".to_string());
 
-		// Initiate the background processors to watch each node.
-		let data_dir = nodes[0].persister.get_data_dir();
-		let callback = move |node: &ChannelManager<InMemorySigner, Arc<ChainMonitor>, Arc<test_utils::TestBroadcaster>, Arc<KeysManager>, Arc<test_utils::TestFeeEstimator>, Arc<test_utils::TestLogger>>| FilesystemPersister::persist_manager(data_dir.clone(), node);
-		let bg_processor = BackgroundProcessor::start(callback, nodes[0].node.clone(), nodes[0].peer_manager.clone(), nodes[0].logger.clone());
-
 		// Go through the channel creation process until each node should have something persisted.
 		let tx = open_channel!(nodes[0], nodes[1], 100000);
 
+		// Initiate the background processors to watch each node.
+		let data_dir = nodes[0].persister.get_data_dir();
+		let persister = move |node: &ChannelManager<InMemorySigner, Arc<ChainMonitor>, Arc<test_utils::TestBroadcaster>, Arc<KeysManager>, Arc<test_utils::TestFeeEstimator>, Arc<test_utils::TestLogger>>| FilesystemPersister::persist_manager(data_dir.clone(), node);
+		let event_handler = |_| {};
+		let bg_processor = BackgroundProcessor::start(persister, event_handler, nodes[0].chain_monitor.clone(), nodes[0].node.clone(), nodes[0].peer_manager.clone(), nodes[0].logger.clone());
```

Comment:
Ah, I see. The comment above `open_channel!` may need updating since it implies that a ChannelManager will be persisted by the time it finishes. 

> Initially, I didn't move it, but the test failed in CI because open_channel! calls get_and_clear_pending_events, which races with BackgroundProcessor. That is it, the event handler may process the FundingGenerationReady event before open_channel! can do so, causing an assertion to fail in open_channel!.
 
Could we add a comment about this? :) Otherwise the test structure is a bit unintuitive to me